<launch>
  <!-- Plane segmentation rosparam namespace-->
  <arg name="namespace"/>

  <!-- Plane segmentation arguments-->
  <arg name="load_table"              default="false" doc="Load the table.pcd located in the load_table_path data folder for convex hull"/>
  <arg name="update_table"            default="true" doc="Update the table data located in the data folder"/>
  <arg name="load_table_path"         default="$(find object_on_table_segmenter)/data/table.pcd" doc="Where the table.pcd located"/>
  <arg name="table_tf"                default="ar_marker_0" doc="Any TF frame located in the table can be used for table segmentation" />
  <arg name="use_tf_surface"          default="false" doc="Use TF surface to define the plane. If false, algorithm will try to find the biggest plane around the surface to get the table convex hull."/>
  <arg name="table_distance_threshold" default="0.02" doc="PCL Plane segmentation distance threshold"/>
  <arg name="table_angular_threshold"  default="2" doc="PCL Plane segmentation angular threshold"/>
  <arg name="table_minimal_inliers"    default="7500" doc="PCL Plane segmentation minimal inliers"/>
  <arg name="use_rosbag"               default="false" doc="Data collection is done on rosbag. Setting this to true should fix the TF frame failed to be found when using rosbag."/>
  <arg name="above_table_min"         default="0.02" doc="Min distance from the table"/>
  <arg name="above_table_max"         default="0.5" doc="Max distance from the table"/>
  <arg name="plane_seg_viewer"        default="false" doc="Enable visualization of the box filtered cloud and the plane segmentation result"/>

  <!-- Plane segmentation parameters -->
  <group ns="$(arg namespace)">
    <param name="load_table"                 type="bool"   value="$(arg load_table)" />
    <param name="update_table"               type="bool"   value="$(arg update_table)" />
    <param name="load_table_path"            type="str"    value="$(arg load_table_path)" />
    <param name="table_tf"                   type="str"    value="$(arg table_tf)"/>
    <param name="table_crop_box_x"           type="double" value="0.5" doc="The x axis distance in meters from the table frame to which the table box should be cropped to, total dimeinsion will be twice this value."/>
    <param name="table_crop_box_y"           type="double" value="0.5" doc="The y axis distance in meters from the table frame to which the table box should be cropped to, total dimeinsion will be twice this value." />
    <param name="table_crop_box_z"           type="double" value="0.5" doc="The z size in meters of the region around which the table box should be defined and cropped to." />
    <param name="tf_frame_based_table_hull_size" type="double" value="0.5" doc="Size in meters from which the table hull will be determined based on the tf frame defined in the table_tf rosparam." />

    <!-- Organized multiplane segmentation parameters. -->
    <!-- See http://docs.pointclouds.org/1.7.0/classpcl_1_1_organized_multi_plane_segmentation.html -->
    <param name="table_distance_threshold"   type="double" value="$(arg table_distance_threshold)"/>
    <param name="table_angular_threshold"    type="double" value="$(arg table_angular_threshold)"/>
    <param name="table_minimal_inliers"      type="double" value="$(arg table_minimal_inliers)"/>

    <param name="use_rosbag"                 type="bool"   value="$(arg use_rosbag)" />
    <param name="use_tf_surface"             type="bool"   value="$(arg use_tf_surface)" />

    <param name="above_table_min"            type="double" value="$(arg above_table_min)" />
    <param name="above_table_max"            type="double" value="$(arg above_table_max)" />
    
    <!-- Plane segmentation parameters -->
    <param name="plane_segmenter_viewer"     type="bool"   value="$(arg plane_seg_viewer)" />
  </group>
</launch>
